txnsync/encodedgroupsmarshalers.go

```diff
@@ -0,0 +1,795 @@
+	stub.BitmaskVersion.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskThreshold.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskSubsigs.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructLsigs(i int, txn *transactions.SignedTxn) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	if txn.Lsig.Logic != nil {
+		if len(stub.BitmaskLogic) == 0 {
+			stub.BitmaskLogic = make(bitmask, bitmaskLen)
+			stub.Logic = make([][]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskLogic.setBit(i)
+		stub.Logic = append(stub.Logic, txn.Lsig.Logic)
+	}
+	if txn.Lsig.Args != nil {
+		if len(stub.BitmaskLogicArgs) == 0 {
+			stub.BitmaskLogicArgs = make(bitmask, bitmaskLen)
+			stub.LogicArgs = make([][][]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskLogicArgs.setBit(i)
+		stub.LogicArgs = append(stub.LogicArgs, txn.Lsig.Args)
+	}
+	if !txn.Lsig.Sig.MsgIsZero() {
+		if len(stub.BitmaskSig) == 0 {
+			stub.BitmaskSig = make(bitmask, bitmaskLen)
+			stub.Sig = make([]byte, 0, int(stub.TotalTransactionsCount)*len(crypto.Signature{}))
+		}
+		stub.BitmaskSig.setBit(i)
+		stub.Sig = append(stub.Sig, txn.Lsig.Sig[:]...)
+	}
+	if txn.Lsig.Msig.Version != 0 {
+		if len(stub.BitmaskVersion) == 0 {
+			stub.BitmaskVersion = make(bitmask, bitmaskLen)
+			stub.Version = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskVersion.setBit(i)
+		stub.Version = append(stub.Version, txn.Lsig.Msig.Version)
+	}
+	if txn.Lsig.Msig.Threshold != 0 {
+		if len(stub.BitmaskThreshold) == 0 {
+			stub.BitmaskThreshold = make(bitmask, bitmaskLen)
+			stub.Threshold = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskThreshold.setBit(i)
+		stub.Threshold = append(stub.Threshold, txn.Lsig.Msig.Threshold)
+	}
+	if txn.Lsig.Msig.Subsigs != nil {
+		if len(stub.BitmaskSubsigs) == 0 {
+			stub.BitmaskSubsigs = make(bitmask, bitmaskLen)
+			stub.Subsigs = make([][]crypto.MultisigSubsig, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskSubsigs.setBit(i)
+		stub.Subsigs = append(stub.Subsigs, txn.Lsig.Msig.Subsigs)
+	}
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructLsigs() {
+	stub.BitmaskLogic.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLogicArgs.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructTransactions(i int, txn *transactions.SignedTxn) error {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	txTypeByte, err := TxTypeToByte(txn.Txn.Type)
+	if err != nil {
+		return fmt.Errorf("failed to deconstructTransactions: %w", err)
+	}
+	if len(stub.BitmaskTxType) == 0 {
+		stub.BitmaskTxType = make(bitmask, bitmaskLen)
+		stub.TxType = make([]byte, 0, int(stub.TotalTransactionsCount))
+	}
+	stub.TxType = append(stub.TxType, txTypeByte)
+	stub.deconstructTxnHeader(i, txn)
+	switch txTypeByte {
```

Comment:
Encoding-decoding of valid transactions should result in the same object. Transactions here are expected to be valid already since they are in the transaction pool.

Reply:
Does the transaction pool check that exactly one set of fields is non-empty? Here is a scenario I'm thinking about. Someone creates a transaction with `Transaction.Type = PaymentTx` and makes `Transaction.KeyregTxnFields` non-empty. Transaction pool accepts it, but peers of this node upon receiving it decode a different transaction that now doesn't pass the signature check. The peers disconnect the sender node.

Even if it's impossible, we are making the correctness of txnsync depend on the evaluator which has very complex logic and which can change. I think it should be avoided.

Reply:
`Does the transaction pool check that exactly one set of fields is non-empty?` hm... yes.
( but indirectly. The txpool is using `WellFormed` that does that )
```golang
func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusParams) error
```